metals/src/main/scala/scala/meta/internal/metals/FoldingRangeProvider.scala

```diff
@@ -0,0 +1,122 @@
+package scala.meta.internal.metals
+
+import java.util
+import java.util.Collections
+import org.eclipse.lsp4j.FoldingRange
+import org.eclipse.lsp4j.FoldingRangeCapabilities
+import scala.meta.Template
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta._
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.FoldingRangeProvider._
+import scala.meta.io.AbsolutePath
+import scala.meta.tokens.Token.KwYield
+
+final class FoldingRangeProvider(val trees: Trees, foldOnlyLines: Boolean) {
+  def getRangedFor(path: AbsolutePath): util.List[FoldingRange] = {
+    trees
+      .get(path)
+      .map(findFoldingRanges)
+      .getOrElse(util.Collections.emptyList())
+  }
+
+  private def findFoldingRanges(tree: Tree): util.List[FoldingRange] = {
+    val ranges = new FoldingRanges(foldOnlyLines)
+
+    tree traverse {
+      case block: Term.Block =>
+        ranges.add(Region, block.pos)
+      case template: Template =>
+        ranges.add(Region, template.pos)
+      case loop: Term.For =>
+        val startLine = loop.pos.startLine
+        val startColumn = loop.pos.startColumn + 3 // just after "for" since there may be no whitespace (e.g. "for{")
+
+        val endLine = loop.body.pos.startLine
+        val endColumn = loop.body.pos.startColumn // must be exact$startColumn, since it can be "}{"
+
+        val range = new FoldingRange(startLine, endLine)
+        range.setStartCharacter(startColumn)
+        range.setEndCharacter(endColumn)
+
+        ranges.add(Region, range)
+
+      // it preserves the whitespaces between "yield" token and the body
```

Comment:
it folds 
```
  def why =
    for {
      x <- ???
    } yield
  
{
    ???
  }
```

to
```
  def why =
    for(...)yield
  
(...)
```

Which does not look the best but - short of circumventing the traversal - cannot be helped.
does it look better this way? probably, but I am not sure
```
  def why =
    for(...)yield(...)
```